src/analysis/interpolation/qgsgridfilewriter.cpp

```diff
@@ -34,95 +36,67 @@ QgsGridFileWriter::QgsGridFileWriter( QgsInterpolator *i, const QString &outputP
 
 int QgsGridFileWriter::writeFile( QgsFeedback *feedback )
 {
-  QFile outputFile( mOutputFilePath );
+  const QFileInfo fi( mOutputFilePath );
+  const QString outputFormat = QgsRasterFileWriter::driverForExtension( fi.suffix() );
+
+  QgsInterpolator::LayerData ld = mInterpolator->layerData().at( 0 );
+  const QgsCoordinateReferenceSystem crs = ld.source->sourceCrs();
 
-  if ( !outputFile.open( QFile::WriteOnly | QIODevice::Truncate ) )
+  std::unique_ptr<QgsRasterFileWriter> writer = std::make_unique<QgsRasterFileWriter>( mOutputFilePath );
+  writer->setOutputProviderKey( QStringLiteral( "gdal" ) );
+  writer->setOutputFormat( outputFormat );
+
+  std::unique_ptr<QgsRasterDataProvider> provider( writer->createOneBandRaster( Qgis::DataType::Float32, mNumColumns, mNumRows, mInterpolationExtent, crs ) );
+  if ( !provider )
   {
+    QgsDebugMsgLevel( QStringLiteral( "Could not create raster output: %1" ).arg( mOutputFilePath ), 2 );
     return 1;
   }
-
-  if ( !mInterpolator )
+  if ( !provider->isValid() )
   {
-    outputFile.remove();
+    QgsDebugMsgLevel( QStringLiteral( "Could not create raster output: %1: %2" ).arg( mOutputFilePath, provider->error().message( QgsErrorMessage::Text ) ), 2 );
     return 2;
   }
 
-  QTextStream outStream( &outputFile );
-#if QT_VERSION < QT_VERSION_CHECK( 6, 0, 0 )
-  outStream.setCodec( "UTF-8" );
-#endif
-  outStream.setRealNumberPrecision( 8 );
-  writeHeader( outStream );
+  provider->setNoDataValue( 1, -9999 );
 
   double currentYValue = mInterpolationExtent.yMaximum() - mCellSizeY / 2.0; //calculate value in the center of the cell
   double currentXValue;
   double interpolatedValue;
 
-  for ( int i = 0; i < mNumRows; ++i )
+  const double step = mNumRows > 0 ? 100.0 / mNumRows : 1;
+  for ( int row = 0; row < mNumRows; row++ )
   {
+    if ( feedback && feedback->isCanceled() )
+    {
+      break;
+    }
+
     currentXValue = mInterpolationExtent.xMinimum() + mCellSizeX / 2.0; //calculate value in the center of the cell
-    for ( int j = 0; j < mNumColumns; ++j )
+
+    QgsRasterBlock block( Qgis::DataType::Float32, mNumColumns, 1 );
+
+    std::vector<float> float32Row( mNumColumns );
```

Comment:
It's be more efficient to move this out of the loop, so that we only incur the allocation cost once.